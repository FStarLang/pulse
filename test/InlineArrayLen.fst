module InlineArrayLen

#lang-pulse
open Pulse

fn basic ()
  returns int
{
  let mut arr = [| 123; 2sz |];
  arr.(0sz);
}

inline_for_extraction noextract
fn gen (x : SizeT.t)
  requires pure (SizeT.v x > 0)
  returns int
{
  let mut arr = [| 123; x |];
  arr.(0sz);
}

fn use ()
  returns int
{
  gen 2sz;
}

inline_for_extraction noextract
fn gen_init (x : SizeT.t)
  (init : unit -> int)
  requires pure (SizeT.v x > 0)
  returns int
{
  let mut arr = [| init(); 2sz |];
  let r = arr.(0sz);
  r
}

fn use_gen_init ()
  returns int
{
  gen_init 2sz (fun _ -> 123);
}

inline_for_extraction noextract
fn gen_init_st (x : SizeT.t)
  (init : unit -> stt int emp (fun _ -> emp))
  requires pure (SizeT.v x > 0)
  returns int
{
  let mut arr = [| init(); 2sz |];
  let r = arr.(0sz);
  r
}

fn use_gen_init_st ()
  returns int
{
  fn init ()
    returns int
  {
    42;
  };
  gen_init_st 2sz init;
}

inline_for_extraction noextract
fn gen_len
  (len : (unit -> x:SizeT.t{SizeT.v x > 0}))
  returns int
{
  let mut arr = [| 123; len() |];
  let r = arr.(0sz);
  r
}

fn use_gen_len ()
  returns int
{
  gen_len (fun _ -> 2sz);
}

inline_for_extraction noextract
fn gen_len_st
  (len : unit -> stt SizeT.t emp (fun r -> pure (SizeT.v r > 0)))
  returns int
{
  let mut arr = [| 123; len () |];
  let r = arr.(0sz);
  r
}

fn use_gen_len_st ()
  returns int
{
  fn len ()
    returns r : SizeT.t
    ensures pure (SizeT.v r > 0)
  {
    42sz;
  };
  gen_len_st len;
}