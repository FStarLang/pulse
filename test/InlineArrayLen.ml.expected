open Prims
let rec (basic : unit -> Prims.int) =
  fun uu___ ->
    let arr =
      Pulse_Lib_Array_Core.alloc (Prims.of_int (123))
        (Stdint.Uint64.of_int (2)) in
    let arr1 =
      Pulse_Lib_Array_Core.op_Array_Access arr Stdint.Uint64.zero () () in
    arr1
let rec (use : unit -> Prims.int) =
  fun uu___ ->
    let arr =
      Pulse_Lib_Array_Core.alloc (Prims.of_int (123))
        (Stdint.Uint64.of_int (2)) in
    let arr1 =
      Pulse_Lib_Array_Core.op_Array_Access arr Stdint.Uint64.zero () () in
    arr1
let rec (use_gen_init : unit -> Prims.int) =
  fun uu___ ->
    let arr =
      Pulse_Lib_Array_Core.alloc (Prims.of_int (123))
        (Stdint.Uint64.of_int (2)) in
    let r = Pulse_Lib_Array_Core.op_Array_Access arr Stdint.Uint64.zero () () in
    let arr1 = r in arr1
let rec (use_gen_init_st : unit -> Prims.int) =
  fun uu___ ->
    let init uu___1 = (Prims.of_int (42)) in
    let __anf = init () in
    let arr = Pulse_Lib_Array_Core.alloc __anf (Stdint.Uint64.of_int (2)) in
    let r = Pulse_Lib_Array_Core.op_Array_Access arr Stdint.Uint64.zero () () in
    let arr1 = r in arr1
let rec (use_gen_len : unit -> Prims.int) =
  fun uu___ ->
    let arr =
      Pulse_Lib_Array_Core.alloc (Prims.of_int (123))
        (Stdint.Uint64.of_int (2)) in
    let r = Pulse_Lib_Array_Core.op_Array_Access arr Stdint.Uint64.zero () () in
    let arr1 = r in arr1
let rec (use_gen_len_st : unit -> Prims.int) =
  fun uu___ ->
    let len uu___1 = (Stdint.Uint64.of_int (42)) in
    let __anf = len () in
    let arr = Pulse_Lib_Array_Core.alloc (Prims.of_int (123)) __anf in
    let r = Pulse_Lib_Array_Core.op_Array_Access arr Stdint.Uint64.zero () () in
    let arr1 = r in arr1
